packages/app-store/zohocalendar/lib/CalendarService.ts

```diff
@@ -0,0 +1,356 @@
+        this.log.error("Error refreshing zoho token", err);
+      }
+      return zohoCredentials;
+    };
+
+    return {
+      getToken: async () => {
+        const isExpired = () => new Date(zohoCredentials.expires_in * 1000).getTime() <= new Date().getTime();
+        return !isExpired() ? Promise.resolve(zohoCredentials) : refreshAccessToken();
+      },
+    };
+  };
+
+  private fetcher = async (endpoint: string, init?: RequestInit | undefined) => {
+    const credentials = await this.auth.getToken();
+
+    return fetch(`https://calendar.zoho.com/api/v1${endpoint}`, {
+      method: "GET",
+      headers: {
+        Authorization: "Bearer " + credentials.access_token,
+        "Content-Type": "application/json",
+        ...init?.headers,
+      },
+      ...init,
+    });
+  };
+
+  private getUserInfo = async () => {
+    const credentials = await this.auth.getToken();
+
+    const response = await fetch(`https://accounts.zoho.com/oauth/user/info`, {
+      method: "GET",
+      headers: {
+        Authorization: "Bearer " + credentials.access_token,
+        "Content-Type": "application/json",
+      },
+    });
+
+    return this.handleData(response, this.log);
+  };
+
+  async createEvent(event: CalendarEvent): Promise<NewCalendarEventType> {
+    let eventId = "";
+    let eventRespData;
+    const calendarId = event.destinationCalendar?.externalId;
+    if (!calendarId) {
+      throw new Error("no calendar id");
+    }
+
+    try {
+      const query = stringify({ eventdata: JSON.stringify(this.translateEvent(event)) });
+
+      const eventResponse = await this.fetcher(`/calendars/${calendarId}/events?${query}`, {
+        method: "POST",
+      });
+      eventRespData = await this.handleData(eventResponse, this.log);
+      eventId = eventRespData.events[0].uid as string;
+    } catch (error) {
+      this.log.error(error);
+      throw error;
+    }
+
+    try {
+      return {
+        ...eventRespData.events[0],
+        uid: eventRespData.events[0].uid as string,
+        id: eventRespData.events[0].id as string,
+        type: "zoho_calendar",
+        password: "",
+        url: "",
+        additionalInfo: {},
+      };
+    } catch (error) {
+      this.log.error(error);
+      await this.deleteEvent(eventId, event, calendarId);
+      throw error;
+    }
+  }
+
+  /**
+   * @param uid
+   * @param event
+   * @returns
+   */
+  async updateEvent(uid: string, event: CalendarEvent, externalCalendarId?: string) {
+    const eventId = uid;
+    let eventRespData;
+    const calendarId = externalCalendarId || event.destinationCalendar?.externalId;
+    if (!calendarId) {
+      this.log.error("no calendar id provided in updateEvent");
+      throw new Error("no calendar id provided in updateEvent");
+    }
+    try {
+      const query = stringify({ eventdata: JSON.stringify(this.translateEvent(event)) });
+
+      const eventResponse = await this.fetcher(`/calendars/${calendarId}/events/${eventId}?${query}`, {
```

Comment:
This endpoint requires an `etag` that we can get by first retrieving the event on the calendar. This is to ensure that the calendar event wasn't updated while this request is being made.